New cross-platform target retroshare-service

G10h4ck opened the pull request:
This PR add a retroshare "pure" service target retroshare-service.
retroshare-service doesn't need any interaction of the user at startup, unlike
retroshare-nogui which requires the user to login on the shell this doesn't
even need a TTY.
At startup this just parse command line, read the PGP key-ring, look for
available locations, and start listening for JSON API requests.
Another difference with retroshare-nogui is that this is capable to
generate/import PGP identities, generate locations, and in general anything
possible through the RetroShare API.
retroshare-service is suitable also to run it as a system service, even
in very constrained systems such as Android or a Docker container.
retroshare-service drop support for libresapi so only the new JSON API
is exposed, it will completely obsolete retroshare-android-service once
retroshare-qml-app is ported to the new JSON API.
A package for Android is also provided to run as a system service independent from which apps will take advantage of it, a button in the menu to eventually stop/restart it is provided too (also because AndroidManifest.xml requires at least an activity to be present ). Different apps could use retroshare-service to provide user level services like blogging, chat, messaging, F2F version of memetro.net etc.
As restbed need full C++11 support a fresh android NDK is needed to compile retroshare-service, in the build_scripts/Android folder a script to create a toolchain based on clang and new NDK is provided.

G10h4ck committed:
Allow CORS in JsonApiServer

G10h4ck committed:
Initial work on a RetroShare cross-platform service

This doesn't need any interacion of the user at startup, unlike
retroshare-nogui which requires the user to login on the shell this doesn't
even need a TTY.
At startup this just parse command line, read the PGP keyring, look for
available locations, and start listening for JSON API requests.
Another difference with retroshare-nogui is that this is capable to
generate/import PGP identities, generate locations, and in general anything
possible through the RetroShare API.
retroshare-service is suitable also to run it as a system service, even
in very constrained systems such as Android ot a Docker container.
retroshare-service drop support for libresapi so only the new JSON API
is exposed, it will completely obsolete retroshare-android-service once
retroshare-qml-app is ported to the new JSON API.

G10h4ck committed:
retroshare-service improve building process

G10h4ck committed:
jsonapi-generator Fix warning that cause errors on mageia

G10h4ck committed:
Fix warning caused by clang defining _GNUC_

G10h4ck committed:
Fix forward declarations class struct missmatch

G10h4ck committed:
Remove overloaded GetRetroshareInvite method

now the original method has default paramether that return invite for
  own node, this way JSON API is more friendly

G10h4ck committed:
Fix deep_search enable/disable at build time

G10h4ck committed:
gxschannel expose blocking create post and channels API

G10h4ck committed:
jsonapi-generator fix param initialization on Android

without this patch Android compilation would fail in cases like
RsPeerId sslid(RsPeerId());

G10h4ck committed:
rstypeserializer fix uninitialized variable usage

G10h4ck committed:
Fix warnings

G10h4ck committed:
retroshare-service now build and runs fine on android (even older ones)

G10h4ck committed:
retroshare-service deep search support for Android

sehraf commented on RetroShare.pro:
```diff
@@ -36,22 +36,34 @@ libretroshare.depends = openpgpsdk libbitdht
     retroshare_nogui.target = retroshare_nogui
+
+    libresapi {
+        retroshare_gui.depends = libresapi
```
Typo, this should be nogui

G10h4ck replied:
great catch!

G10h4ck committed:
Fix compilation due to RsPeers::GetRetroshareInvite

G10h4ck committed:
Fix compilation due to .pro copy/paste error

Thanks sehraf for spotting this that was driving me mad!

G10h4ck committed:
Fix type in .pri breaking non-cross-compiling

G10h4ck committed:
Add OBS packaging recipes for retroshare-service

G10h4ck committed:
Expose JSON API for file filtering

G10h4ck committed:
Expose JSON API for ForceDirectoryCheck

G10h4ck committed:
jsonwrappersincl target always fails if doxygen fail

G10h4ck committed:
retroshare-service add missing doxygen dependency on OBS

G10h4ck committed:
Fix bug in JSON t_RsFlags32 deserialization

If the member was not present in the JSON it was zeroed while it should
be left untouched in that case.
This was discovered because in JSON API defualt argument value for
methods wasn't onored for t_RsFlags32 arguments.

G10h4ck committed:
Expose /rsFiles/requestDirDetails via JSON API

This method is useful to browse own shared directries, or directries
shared by friends whith browseable permission

G10h4ck committed:
Implement mostly decent JSON API for channels and post creation

Now the possible id of the cannel/post is returned and channel/post is
indexed on creation without having to wait for gxsutil scan to happen

Note that the id is correct in normal usage, but if one try to overload
the API per eventually get the id of another channel/post with same data
but created by a different call, it is not dangerous so one trying to
overload the API is just getting a possibly wrong but equivalent
channel/post id returned.

G10h4ck committed:
Merge branch 'master' into jsonapi

G10h4ck committed:
Suppress aliasing warning in DirDetails and DirStub

Order of serialization doesn't matter ATM as we use those only for
  JSON API.
Those classes should be rewritten anyway with proper naming of the
  memebers and using std::uintptr_t instead of void* and rstime_t for
  timestamps

G10h4ck committed:
Re-add ftServer::requestDirDetails implementation

The implementation has disappeared merging master causing linking error

csoler commented on jsonapi-generator/src/jsonapi-generator.cpp:
```diff
@@ -45,7 +45,7 @@ int main(int argc, char *argv[])
 	{
 		qDebug() << "Usage: jsonapi-generator SOURCE_PATH OUTPUT_PATH";
-		return EINVAL;
+		return -EINVAL;
```
I'm surprised by the ngative value here. Is there some kind of hidden protocol to interpret that value elsewhere? If so, it could be documented.

G10h4ck replied:
It is common practice to return -Enumber in case of error in executable

csoler commented on libretroshare/src/jsonapi/jsonapi.cpp:
```diff
@@ -41,6 +41,26 @@
+/*static*/ const std::multimap<std::string, std::string>
+JsonApiServer::corsHeaders =
+{
+    { "Access-Control-Allow-Origin", "*" },
```
does this line mean you can access the API from another IP than 127.0.0.1 ? If so, it's terribly insecure

G10h4ck replied:
No, it doesn't mean that, it means that if you access the API trough a browser it doesn't mind the domain name you have on your urlbar, so for example if an APP based on electron or other web framework uses RS API it could access even if they define their own alias of 127.0.0.1 to be more user friendly or use another port fro their own HTTP server on same machine. RetroShare JSON API bind on 127.0.0.1:9092 if it is not otherwise specified by command-line options, or by RetroShare GUI options.

csoler commented on libretroshare/src/retroshare/rsfiles.h:
```diff
@@ -182,13 +183,22 @@ class FileTree
+	virtual void serial_process(RsGenericSerializer::SerializeJob j,
+	                            RsGenericSerializer::SerializeContext& ctx)
+	{
+		RS_SERIAL_PROCESS(mFilename);
```
This notation is very cmpact, but it has a drawback: the variables j and ctx cannot have a different name. Not really a problem, but if we document the serialization that will need to be said somewhere.

G10h4ck replied:
If I remember well it is documented on the header where that macro is defined (anyway this has been introduced lot of time ago so it is not strictly related to this PR ;) )

csoler commented on libretroshare/src/services/p3gxschannels.cc:
```diff
@@ -1036,6 +1042,108 @@ bool p3GxsChannels::getChannelsContent(
 	return getPostData(token, posts, comments);
 }
 
+bool p3GxsChannels::createChannel(RsGxsChannelGroup& channel)
```
The async function to do that is called createGroup(), as for all other GXS services. Naming it createChannel is of course more appropriate but it also hides the fact that the right function to call is the other one. I suggest to name this one createChannel_synchroneous() or somthing like this.

G10h4ck replied:
I don't think one is wrong and the other is right, they offer different behavior, I kept the other with the original name because it have the asynchronous semantic of GXS style API like all other services. The documentation in the header clearly states that that function is blocking, also I would avoid the longer name.

csoler commented on libretroshare/src/services/p3gxschannels.cc:
```diff
@@ -1036,6 +1042,108 @@ bool p3GxsChannels::getChannelsContent(
+	time_t endCreation = time(nullptr);
+
+	std::list<RsGroupMetaData> channels;
+	if(!getChannelsSummaries(channels)) return false;
```
There is no guaranty that this call will return the newly created channel, since the wait() above only waits until the creation order is put into some queue.

G10h4ck replied:
Nope, I have dig into the GXS code and the token is updated correctly only once the channel is effectively created, published and stored into the database, so we get to this point only if the channel has been created successfully.

csoler commented on libretroshare/src/services/p3gxschannels.cc:
```diff
@@ -1036,6 +1042,108 @@ bool p3GxsChannels::getChannelsContent(
+	bool found = false;
+	for(const RsGroupMetaData& chan : channels)
+	{
+		if( IS_GROUP_ADMIN(chan.mSubscribeFlags)
```
Indeed, that is really ugly. Another way to go would be to add the creation token into RsGroupMetaData, or use some kind of preliminary identifier, returned by createGroup() that can further be used to retrieve the correct channel.

csoler replied:
Besides if the channel creation fails for some reason, a function calling getChannelSummary() indefinitly would never return.

G10h4ck replied:
> Indeed, that is really ugly. Another way to go would be to add the creation token into RsGroupMetaData, or use some kind of preliminary identifier, returned by createGroup() that can further be used to retrieve the correct channel.

I have experimented various ways of doing that, the problem is that while we keep track of token and token status in RsGxsDataAccess also for channel creation like we do for data read request, RsGxsDataAccess cannot keep track of channel creation data and request because it is done into RsGenExchange so triyng to keep some data into RsGxsDataAccess about the channel a race condition between RsGxsDataAccess and RsGenExchange raise and the channel is never created (also RS need to be stopped by kill -9 in that case...)

The most elegant working solution I could think of after many experiments is to move the channel creation into RsGxsDataAccess and fully keep track of it there like we do with the other GxsRequests, but indeed attempting to to that caused a refactoring chain reaction, and I have no time for that right now and I think it should eventually done into it's own pull request

> use some kind of preliminary identifier

token is meant for that use but is not usable to retrieve data in this case because of what I explained before

G10h4ck replied:
> Besides if the channel creation fails for some reason, a function calling getChannelSummary() indefinitly would never return.

Nope it would return false in that case quite fast, anyway in this context we would never call that function if channel creation failed because if `waitToken` return something different than `COMPLETE` `createChannel` return false before of getting there.

Also take in account that `waitToken` by default has a timeout of 500ms

csoler commented on libretroshare/src/services/p3gxschannels.cc:
```diff
@@ -1036,6 +1042,108 @@ bool p3GxsChannels::getChannelsContent(
+	return found;
+}
+
+bool p3GxsChannels::createPost(RsGxsChannelPost& post)
```
Same comments for this function.

csoler commented on retroshare-gui/src/gui/GetStartedDialog.cpp:
```diff
@@ -236,7 +236,7 @@ void GetStartedDialog::inviteFriends()
 	{
 		RsAutoUpdatePage::lockAllEvents();
 
-		cert = rsPeers->GetRetroshareInvite(false);
```
I put (false,false)  on purpose in the HomePage: to avoid leaking previous IPs unless the user really wants this. I think we should do the same everywhere. If I remember correctly, the false should only be removed on android because it's a better compromise on android to leak as many ips as possible.

G10h4ck replied:
The false there was applied to `includeSignatures` as you can see here https://github.com/RetroShare/RetroShare/blob/18eee374f1b56ad41760c8ec17f2e90e6eb4717c/libretroshare/src/retroshare/rspeers.h#L557

In this PR I have unified the two versions of `GetRetroshareInvite` so this change is only to fix compilation error but doesn't change the semantic as thi sother function would be called now
https://github.com/RetroShare/RetroShare/pull/1363/files#diff-0a044481cf0cc5773dc30a3a6aeffee8R537
About not exporting the last recently used 10 IPv# i am not sure about how much it improves the privacy or just make connection more difficult, but I think as that is a semantic change should be part of another PR/discussion ;)

csoler commented on retroshare-gui/src/gui/connect/ConnectFriendWizard.cpp:
```diff
@@ -127,7 +127,7 @@ ConnectFriendWizard::ConnectFriendWizard(QWidget *parent) :
     body = GetStartedDialog::GetInviteText();
 	
     body += "\n" + GetStartedDialog::GetCutBelowText();
-    body += "\n\n" + QString::fromUtf8(rsPeers->GetRetroshareInvite(false).c_str());
```
Same comment here about not leaking all IPs -> (false,false)

csoler commented on retroshare-service/src/retroshare-service.cc:
```diff
@@ -0,0 +1,78 @@
+		argv2[argc+1] = val;
+		RsInit::InitRetroShare(argc2, argv2, true);
+	}
+	else RsInit::InitRetroShare(argc, argv, true);
```
I presume it's not currently possible to run an auto-configured Tor node with this.

G10h4ck replied:
Isn't it supported by  `RsInit::InitRetroShare(...)` ?
This snippet of code only add `--jsonApiPort` if it is not already specified because JSON API is the only mean of interaction available on retroshare-service, but won't alter other command line options

csoler reviewed (commented):
Congrats for this huge amount of work! I left a few comments.

G10h4ck committed:
Fix crash in p3Peers::loadCertificateFromString

This function is part of the public API so it must be safe to call with
any input, before this commit if would crash if feeded with a broken or
empty certificate radix string.
